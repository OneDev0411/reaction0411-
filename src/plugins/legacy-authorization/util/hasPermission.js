import { createRequire } from "module";
import Logger from "@reactioncommerce/logger";
import ReactionError from "@reactioncommerce/reaction-error";

const require = createRequire(import.meta.url); // eslint-disable-line

const { curryN } = require("ramda");

const GLOBAL_GROUP = "__global_roles__";

/**
 * @name hasPermission
 * @param {Object} context App context
 * @param {Object} resource - resource user is trying to access
 * @param {Object} action - action user is trying to perform to be passed to in the GQL query
 * @param {Object} authContext - context data to verify permissions against
 * @param {String} [authContext.owner] - The owner of the resource requested
 * @param {String} [authContext.shopId] - The shop ID for which the permissions are needed. If not set,
 *   only global roles will be checked.
 * @param {Array} [authContext.legacyRoles] - TEMPORARY: roles that match up with the legacy roles package
 * @returns {Boolean} - true/false
 */
export default async function hasPermission(context, resource, action, authContext) {
  const { user } = context;

  // If the current user is the owner of a resource we are trying to check,
  // such as an order or data on a user profile, they are authorized to perform the action
  if (authContext && authContext.owner && authContext.owner === context.userId) return true;

  if (!user || !user.roles) return false;

  if (!resource) throw new ReactionError("invalid-param", "Resource must be provided");

  if (!action) throw new ReactionError("invalid-param", "Action must be provided");

  if (!authContext) throw new ReactionError("invalid-param", "authContext must be provided");

  if (!authContext.shopId) throw new ReactionError("invalid-param", "shopId (roleGroup) must be provided");

  const { legacyRoles: permissions, shopId: roleGroup } = authContext; // TODO(pod-auth): temporarily provide legacy roles

  if (!Array.isArray(permissions)) throw new ReactionError("invalid-param", "permissions must be an array of strings");
  if (roleGroup !== undefined && roleGroup !== null && (typeof roleGroup !== "string" || roleGroup.length === 0)) {
    throw new ReactionError("invalid-param", "roleGroup must be a non-empty string");
  }

  const checkRoles = permissions.slice(0);

  // This should always return true for owners
  if (!checkRoles.includes("owner")) checkRoles.push("owner");

  const { roles } = user;

  // always check GLOBAL_GROUP
  const globalRoles = roles[GLOBAL_GROUP];
  if (Array.isArray(globalRoles) && checkRoles.some((role) => globalRoles.includes(role))) return true;

  if (roleGroup) {
    // convert any periods to underscores for MongoDB compatibility
    const group = roleGroup.replace(/\./g, "_");

    const groupRoles = roles[group];

    if (Array.isArray(groupRoles) && checkRoles.some((role) => groupRoles.includes(role))) return true;
  }

  Logger.debug(`User ${user._id} has none of [${checkRoles.join(", ")}] permissions`);

  return false;
}

const hasPermissionCurried = curryN(2, hasPermission);

/**
 * @summary Get a `hasPermission` function bound to the current user context
 * @param {Object} context App context
 * @return {Function} hasPermission function for `context`
 */
export function getHasPermissionFunctionForUser(context) {
  return hasPermissionCurried(context);
}
