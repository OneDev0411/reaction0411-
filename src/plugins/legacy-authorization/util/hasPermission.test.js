import hasPermission from "./hasPermission.js";

test("returns false if no user", async () => {
  const result = await hasPermission({ user: null }, "resource", "action", { shopId: "scope" });
  expect(result).toBe(false);
});

test("returns false if no user.roles", async () => {
  const result = await hasPermission({ user: { roles: null } }, "resource", "action", { shopId: "scope" });
  expect(result).toBe(false);
});

test("throws if no resource", async () => {
  const result = hasPermission(
    {
      user: {
        roles: {
          __global_roles__: ["organization:system:entity/action"], // eslint-disable-line camelcase
          scope: ["can_eat"]
        }
      }
    },
    null,
    "action",
    { shopId: "scope" }
  );
  expect(result).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if no action", async () => {
  const result = hasPermission(
    {
      user: {
        roles: {
          __global_roles__: ["organization:system:entity/action"], // eslint-disable-line camelcase
          scope: ["can_eat"]
        }
      }
    },
    "organization:system:entity",
    null,
    { shopId: "scope" }
  );
  expect(result).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if no authContext", async () => {
  const result = hasPermission(
    {
      user: {
        roles: {
          __global_roles__: ["organization:system:entity/action"], // eslint-disable-line camelcase
          scope: ["can_eat"]
        }
      }
    },
    "organization:system:entity",
    "action",
    null
  );
  expect(result).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if shopId is present but not a string", async () => {
  const result = hasPermission(
    {
      user: {
        roles: {
          __global_roles__: ["organization:system:entity/action"], // eslint-disable-line camelcase
          scope: ["can_eat"]
        }
      }
    },
    "organization:system:entity",
    "action",
    {
      shopId: ["thisIsNotAString"]
    }
  );
  expect(result).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if shopId is present but an empty string", async () => {
  const result = hasPermission(
    {
      user: {
        roles: {
          __global_roles__: ["organization:system:entity/action"], // eslint-disable-line camelcase
          scope: ["can_eat"]
        }
      }
    },
    "organization:system:entity",
    "action",
    {
      shopId: ""
    }
  );
  expect(result).rejects.toThrowErrorMatchingSnapshot();
});

test("returns true if in global role, even if not in group-scope role", async () => {
  const result = await hasPermission(
    {
      user: {
        roles: {
          __global_roles__: ["organization:system:entity/action"], // eslint-disable-line camelcase
          scope: ["otherOrganization:otherSystem:otherEntity/otherAction"]
        }
      }
    },
    "organization:system:entity",
    "action",
    {
      shopId: "SHOP_ID"
    }
  );
  expect(result).toBe(true);
});

test("returns true if in group-scope role but not in global role", async () => {
  const result = await hasPermission(
    {
      user: {
        roles: {
          __global_roles__: ["otherOrganization:otherSystem:otherEntity/otherAction"], // eslint-disable-line camelcase
          scope: ["organization:system:entity/action"]
        }
      }
    },
    "organization:system:entity",
    "action",
    {
      shopId: "scope"
    }
  );
  expect(result).toBe(true);
});

test("returns true if in role in both scopes", async () => {
  const result = await hasPermission(
    {
      user: {
        roles: {
          __global_roles__: ["organization:system:entity/action"], // eslint-disable-line camelcase
          scope: ["organization:system:entity/action"]
        }
      }
    },
    "organization:system:entity",
    "action",
    {
      shopId: "scope"
    }
  );
  expect(result).toBe(true);
});

test("returns false if not in any role in either scope", async () => {
  const result = await hasPermission(
    {
      user: {
        roles: {
          __global_roles__: ["otherOrganization:otherSystem:otherEntity/otherAction"], // eslint-disable-line camelcase
          scope: ["can_eat"]
        }
      }
    },
    "organization:system:entity",
    "action",
    {
      shopId: "scope"
    }
  );
  expect(result).toBe(false);
});

test("returns true if has owner role, even if not explicitly in the permissions array", async () => {
  const result = await hasPermission(
    {
      user: {
        roles: {
          __global_roles__: ["otherOrganization:otherSystem:otherEntity/otherAction"], // eslint-disable-line camelcase
          scope: ["reaction:legacy:shops/owner"]
        }
      }
    },
    "organization:system:entity",
    "action",
    {
      shopId: "scope"
    }
  );
  expect(result).toBe(true);
});
