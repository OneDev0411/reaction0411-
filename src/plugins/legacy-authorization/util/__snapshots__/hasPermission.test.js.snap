// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws if no action 1`] = `"Action must be provided"`;

exports[`throws if no authContext 1`] = `"authContext must be provided"`;

exports[`throws if no resource 1`] = `"Resource must be provided"`;

exports[`throws if shopId is present but an empty string 1`] = `"shopId must be a non-empty string"`;

exports[`throws if shopId is present but not a string 1`] = `"shopId must be a non-empty string"`;
